demos/custom-tabs-example-app/src/main/java/org/chromium/customtabsdemos/EngagementSignalsActivity.java

```diff
@@ -0,0 +1,190 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package org.chromium.customtabsdemos;
+
+import static androidx.browser.customtabs.CustomTabsIntent.ACTIVITY_HEIGHT_ADJUSTABLE;
+
+import android.net.Uri;
+import android.os.Bundle;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AppCompatActivity;
+
+import android.os.RemoteException;
+import android.util.Log;
+import android.view.View;
+import android.widget.EditText;
+import android.widget.TextView;
+
+import androidx.browser.customtabs.CustomTabsCallback;
+import androidx.browser.customtabs.CustomTabsClient;
+import androidx.browser.customtabs.CustomTabsIntent;
+import androidx.browser.customtabs.CustomTabsSession;
+import androidx.browser.customtabs.EngagementSignalsCallback;
+
+public class EngagementSignalsActivity extends AppCompatActivity implements View.OnClickListener {
+    private static final String TAG = EngagementSignalsActivity.class.getSimpleName();
+    private static final int INITIAL_HEIGHT_DEFAULT_PX = 600;
+
+    private EditText mUrlEditText;
+    private TextView mTextVerticalScroll;
+    private TextView mTextGreatestPercentage;
+    private TextView mTextSessionEnd;
+    private TextView mTextNavigation;
+
+    private ServiceConnection mConnection;
+    private CustomTabsClient mClient;
+    private CustomTabsSession mCustomTabsSession;
+
+    private final ServiceConnectionCallback mServiceConnectionCallback = new ServiceConnectionCallback() {
+        @Override
+        public void onServiceConnected(CustomTabsClient client) {
+            mClient = client;
+            mCustomTabsSession = mClient.newSession(mCustomTabsCallback);
+            try {
+                boolean engagementSignalsApiAvailable = mCustomTabsSession.isEngagementSignalsApiAvailable(Bundle.EMPTY);
+                if (!engagementSignalsApiAvailable) {
+                    Log.d(TAG, "CustomTab Engagement signals not available, make sure to use the " +
+                            "latest Chrome version and enable via chrome://flags/#cct-real-time-engagement-signals");
+                    return;
+                }
+                mCustomTabsSession.setEngagementSignalsCallback(mEngagementSignalsCallback, Bundle.EMPTY);
```

Comment:
#setEngagementSignalsCallback can return false if the callback cannot be set so I think we should handle that scenario as well.

Reply:
What's the difference to `isEngagementSignalsApiAvailable` returning false?

Reply:
To be fair, not really a difference as, at least on the Chrome side, currently we are calling the same logic internally to check if this can be set or not. However, things could have happened in between these 2 calls and the setting might have not happened. 

Given that the set* API returns a value, maybe we could handle the failure case separately and inform that the callback was not set? Not sure if this is the direction you want to go, but maybe the de demo app we could have a button that allows the user of the app to set the engagement signals callback/logic and start using it and one to "unset" it and stop using it and once the ES are successfully set the set button becomes disabled and unset becomes enabled. For onSessionEnded/not being able to set the ES we could reset the button status and inform the user of what happened. I wonder if something similar to this would showcase more all the scenario a developer might need to handle.